### PR TITLE
Add w8a8 quantized matmul kernel

### PR DESCRIPTION
**Pull request by vanbasten23**
This PR adds:
- a Pallas kernel for doing w8a8 quantized matmul
- tuned table for v6e

Test plan
- pip install pytest
- pytest pytorch/xla/test/test_quantized_matmul_pallas_kernel.py -s

### CODE REVIEW

File: torch_xla/experimental/pallas_kernels/quantized_matmul_kernel.py [hunk 1]
```diff
@@ -0,0 +1,321 @@
...
+        'out_block_size',
+        'in_block_size',
+    ])
+def quantized_matmul(
+    x: jax.Array,  # [bs, n_input_features]
+    w: jax.Array,  # [n_output_features, n_input_features]
+    scalar: jax.Array,  # [n_output_features]
+    zero_point: jax.Array | None = None,
+    block_size: jax.Array | None = None,  # i32[1]
+    int4_weight: bool = False,
```
Comment (yaochengji) [inline torch_xla/experimental/pallas_kernels/quantized_matmul_kernel.py]: Can we get this from w.dtype?

File: torch_xla/experimental/pallas_kernels/quantized_matmul_kernel.py [hunk 2]
```diff
@@ -0,0 +1,321 @@
...
+    batch_block_size: int | None = None,
+    out_block_size: int | None = None,
+    in_block_size: int | None = None,
+    vmem_limit_bytes: int | None = 64 * 1024 * 1024,
+):
+  assert zero_point is None, "Not implemented: zero_point is not supported."
+  assert block_size is None, "Not implemented: block_size is not supported."
+  assert not int4_weight, "Not implemented: int4_weight is not supported."
+
+  # x_max_val cannot be [bs, 128] because it'll be costly to send
```
Comment (yaochengji) [inline torch_xla/experimental/pallas_kernels/quantized_matmul_kernel.py]: Why did you say 128 here?

File: torch_xla/experimental/pallas_kernels/quantized_matmul_kernel.py [hunk 3]
```diff
@@ -0,0 +1,321 @@
...
+  if orig_bs < padded_bs:
+    x = jnp.pad(x, ((0, padded_bs - orig_bs), (0, 0)))
+    x_max_val = jnp.pad(x_max_val, ((0, 0), (0, padded_bs - orig_bs)))
+  padded_out_features = _next_multiple(orig_out_features, out_block_size)
+  if orig_out_features < padded_out_features:
+    w = jnp.pad(w, ((0, padded_out_features - orig_out_features), (0, 0)))
+    scalar = jnp.pad(scalar, (0, padded_out_features - orig_out_features))
+  padded_in_features = _next_multiple(orig_in_features, in_block_size)
+  if orig_in_features < padded_in_features:
+    x = jnp.pad(x, ((0, 0), (0, padded_in_features - orig_in_features)))
```
Comment (yaochengji) [inline torch_xla/experimental/pallas_kernels/quantized_matmul_kernel.py:139]: Do you think pad is quite expansive? Can we use mask in kernel?

File: torch_xla/experimental/pallas_kernels/quantized_matmul_kernel.py [hunk 4]
```diff
@@ -0,0 +1,321 @@
...
+        'int4_weight',
+        'quantize_activation',
+        'batch_block_size',
+        'out_block_size',
+        'in_block_size',
+    ])
+def quantized_matmul(
+    x: jax.Array,  # [bs, n_input_features]
+    w: jax.Array,  # [n_output_features, n_input_features]
+    scalar: jax.Array,  # [n_output_features]
```
Comment (yaochengji) [inline torch_xla/experimental/pallas_kernels/quantized_matmul_kernel.py:101]: rename to `bias`?

File: test/test_quantized_matmul_pallas_kernel.py [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+        x,
+        q_w,
+        scalar_w,
+        quantize_activation=quantize_activation,
+        batch_block_size=batch_block_size,
+        out_block_size=out_block_size,
+        in_block_size=in_block_size).block_until_ready()
+    expected = jax.lax.dot_general(
+        x_copy, w_copy, dimension_numbers=(((1,), (1,)), ((), ())))
+    # print(
```
Comment (yaochengji) [inline test/test_quantized_matmul_pallas_kernel.py]: Remove the print?

File: torch_xla/experimental/pallas_kernels/quantized_matmul_kernel.py [hunk 6]
```diff
@@ -0,0 +1,321 @@
...
+def matmul_kernel(
+    x_ref,  # (batch_block_size, in_block_size)
+    w_ref,  # (out_block_size, in_block_size)
+    scalar_ref,  # (1, out_block_size)
+    x_max_val,  # (1, batch_block_size)
+    out_ref,  # (batch_block_size, out_block_size)
+    acc_ref,  # (batch_block_size, out_block_size)
+    *,
+    quantize_activation,
+    batch_block_size,
```
Comment (yaochengji) [inline torch_xla/experimental/pallas_kernels/quantized_matmul_kernel.py:36]: I think usually we use the names: m, n, k.

File: torch_xla/experimental/pallas_kernels/quantized_matmul_kernel.py [hunk 7]
```diff
@@ -0,0 +1,321 @@
...
+
+def matmul_kernel(
+    x_ref,  # (batch_block_size, in_block_size)
+    w_ref,  # (out_block_size, in_block_size)
+    scalar_ref,  # (1, out_block_size)
+    x_max_val,  # (1, batch_block_size)
+    out_ref,  # (batch_block_size, out_block_size)
+    acc_ref,  # (batch_block_size, out_block_size)
+    *,
+    quantize_activation,
```
Comment (yaochengji) [inline torch_xla/experimental/pallas_kernels/quantized_matmul_kernel.py:35]: Seems only int8 quantization is supported at the moment. Could you clarify it in the function name?

File: torch_xla/experimental/pallas_kernels/quantized_matmul_kernel.py [hunk 8]
```diff
@@ -0,0 +1,321 @@
...
+        'quantize_activation',
+        'batch_block_size',
+        'out_block_size',
+        'in_block_size',
+    ])
+def quantized_matmul(
+    x: jax.Array,  # [bs, n_input_features]
+    w: jax.Array,  # [n_output_features, n_input_features]
+    scalar: jax.Array,  # [n_output_features]
+    zero_point: jax.Array | None = None,
```
Comment (yaochengji) [inline torch_xla/experimental/pallas_kernels/quantized_matmul_kernel.py:102]: I think zero_point should be computed from x if we want to use asymmetric quantization?

File: torch_xla/experimental/pallas_kernels/quantized_matmul_kernel.py (hunk 1 above)
Comment (vanbasten23) [reply torch_xla/experimental/pallas_kernels/quantized_matmul_kernel.py]: I copied the interface from https://github.com/pytorch/xla/blob/c71ff48bb8d758bef57835bc2814353d74fbfb3e/torch_xla/experimental/xla_quantized_matmul.py#L78-L84.

But I think it makes sense to get it from w.dtype so I removed it.

File: torch_xla/experimental/pallas_kernels/quantized_matmul_kernel.py (hunk 2 above)
Comment (vanbasten23) [reply torch_xla/experimental/pallas_kernels/quantized_matmul_kernel.py]: 128 is the minormost dimension of the vreg. I revised the comment to 
```
# x_max_val cannot be [bs, 128] where 128 is the minormost dimension of the vreg because it'll be costly to store
# [bs_block_size, 128] in VMEM ([bs_balock_size, 1:] will be padding).
```

File: torch_xla/experimental/pallas_kernels/quantized_matmul_kernel.py (hunk 3 above)
Comment (vanbasten23) [reply torch_xla/experimental/pallas_kernels/quantized_matmul_kernel.py:139]: For us, the case where padding is needed is when the dimension is less than the block sizes (note, the 3 block sizes need to be at least 128). So when batch_size = 16, then we need to pad anyway and cannot do mask..

File: torch_xla/experimental/pallas_kernels/quantized_matmul_kernel.py (hunk 4 above)
Comment (vanbasten23) [reply torch_xla/experimental/pallas_kernels/quantized_matmul_kernel.py:101]: this is the weight scalar in the affine transformation. We don't use bias here or in the torch_xla version of the quantized matmul https://github.com/pytorch/xla/blob/c71ff48bb8d758bef57835bc2814353d74fbfb3e/torch_xla/experimental/xla_quantized_matmul.py#L78-L84

File: test/test_quantized_matmul_pallas_kernel.py (hunk 5 above)
Comment (vanbasten23) [reply test/test_quantized_matmul_pallas_kernel.py]: done

File: torch_xla/experimental/pallas_kernels/quantized_matmul_kernel.py (hunk 6 above)
Comment (vanbasten23) [reply torch_xla/experimental/pallas_kernels/quantized_matmul_kernel.py:36]: I did that initially but later I got confused sometimes when I map them to bs, n_input_features, n_output_features. So I prefer to keep that way

File: torch_xla/experimental/pallas_kernels/quantized_matmul_kernel.py (hunk 7 above)
Comment (vanbasten23) [reply torch_xla/experimental/pallas_kernels/quantized_matmul_kernel.py:35]: w8a16 will be the next step: quantize_activation=False. Can I keep the function name and add the w8a16 support in the later PR?

File: torch_xla/experimental/pallas_kernels/quantized_matmul_kernel.py (hunk 8 above)
Comment (vanbasten23) [reply torch_xla/experimental/pallas_kernels/quantized_matmul_kernel.py:102]: You're right. But this is the zero_point of the weight instead of the activation https://github.com/pytorch/xla/blob/c71ff48bb8d758bef57835bc2814353d74fbfb3e/torch_xla/experimental/xla_quantized_matmul.py#L95C7-L97. When we quantize the activation, I used the symmetric quantization same as torch_xla https://github.com/pytorch/xla/blob/c71ff48bb8d758bef57835bc2814353d74fbfb3e/torch_xla/experimental/xla_quantized_matmul.py#L56-L74

File: torch_xla/experimental/pallas_kernels/quantized_matmul_kernel.py [hunk 9]
```diff
@@ -0,0 +1,321 @@
...
+import jax
+from jax import lax
+from jax.experimental import pallas as pl
+from jax.experimental.pallas import tpu as pltpu
+import jax.numpy as jnp
+
+
+def _quantize_array(
+    x: jax.Array,  # [bs_block_size, in_block_size]
+    x_max_val: jax.Array,  # [1, bs_block_size]
```
Comment (lsy323) [inline torch_xla/experimental/pallas_kernels/quantized_matmul_kernel.py]: maybe rename to `x_abs_max_val` to indicate it's abs max

File: torch_xla/experimental/pallas_kernels/quantized_matmul_kernel.py [hunk 10]
```diff
@@ -0,0 +1,321 @@
...
+    n_bits: int = 8,
+):
+  int_min = -2**(n_bits - 1)
+  int_max = 2**(n_bits - 1) - 1
+  scale = (x_max_val / int_max).T  # [bs_block_size, 1]
+  # Need to explicitly cast to f32 because Mosaic can't directly jnp.round a
+  # bf16 array.
+  # It seems x/0 in Pallas generates inf/-inf instead of an exception.
+  x_int = jnp.clip(
+      jnp.round((x / scale).astype(jnp.float32)), int_min,
```
Comment (lsy323) [inline torch_xla/experimental/pallas_kernels/quantized_matmul_kernel.py:22]: May I ask why do we have `.astype(jnp.float32)` here? It doesn't seem to improve the precision

File: torch_xla/experimental/pallas_kernels/quantized_matmul_kernel.py [hunk 11]
```diff
@@ -0,0 +1,321 @@
...
+      int_max).astype(jnp.int8)
```
Comment (lsy323) [inline torch_xla/experimental/pallas_kernels/quantized_matmul_kernel.py:23]: The function is hardcoded to quantized tensors to int8, so I guess we don't need `n_bits` for now?

File: torch_xla/experimental/pallas_kernels/quantized_matmul_kernel.py [hunk 12]
```diff
@@ -0,0 +1,319 @@
...
+          vmem_limit_bytes=vmem_limit_bytes,
+      ),
+  )
+
+  out = kernel(x, w, scalar, x_max_val)
+
+  return out[:orig_bs, :orig_out_features]
+
+
+def quantize_array(x, n_bits: int = 8, dim: int = -1):
```
Comment (lsy323) [inline torch_xla/experimental/pallas_kernels/quantized_matmul_kernel.py]: Is this redundant? I see a `_quantize_array` above

File: torch_xla/experimental/pallas_kernels/quantized_matmul_kernel.py [hunk 13]
```diff
@@ -0,0 +1,319 @@
...
+        'out_block_size',
+        'in_block_size',
+        'vmem_limit_bytes',
+    ])
+def quantized_matmul(
+    x: jax.Array,  # [bs, n_input_features]
+    w: jax.Array,  # [n_output_features, n_input_features]
+    scalar: jax.Array,  # [n_output_features]
+    zero_point: jax.Array | None = None,
+    block_size: jax.Array | None = None,  # i32[1]
```
Comment (lsy323) [inline torch_xla/experimental/pallas_kernels/quantized_matmul_kernel.py]: nit: Maybe add a comment on this `block_size` arg, it's the block size for blockwise quantization, not the block size in pallas.

Comment (lsy323) [reply torch_xla/experimental/pallas_kernels/quantized_matmul_kernel.py]: Or we rename to `quant_block_size`

File: torch_xla/experimental/pallas_kernels/quantized_matmul_kernel.py (hunk 7 above)
Comment (yaochengji) [reply torch_xla/experimental/pallas_kernels/quantized_matmul_kernel.py:35]: Yeah, but I'm considering int8 vs. fp8.

Comment (vanbasten23) [reply torch_xla/experimental/pallas_kernels/quantized_matmul_kernel.py:35]: Got it. Renamed to quantized_matmul_int8

File: torch_xla/experimental/pallas_kernels/quantized_matmul_kernel.py (hunk 9 above)
Comment (vanbasten23) [reply torch_xla/experimental/pallas_kernels/quantized_matmul_kernel.py]: done

File: torch_xla/experimental/pallas_kernels/quantized_matmul_kernel.py (hunk 10 above)
Comment (vanbasten23) [reply torch_xla/experimental/pallas_kernels/quantized_matmul_kernel.py:22]: Yeah, it's because there is a Mosaic limitation when it jnp.round a bf16 type. So I have to cast it to f32 to work around the limitation.

File: torch_xla/experimental/pallas_kernels/quantized_matmul_kernel.py (hunk 12 above)
Comment (vanbasten23) [reply torch_xla/experimental/pallas_kernels/quantized_matmul_kernel.py]: removed it from the file.

File: torch_xla/experimental/pallas_kernels/quantized_matmul_kernel.py (hunk 13 above)
Comment (vanbasten23) [reply torch_xla/experimental/pallas_kernels/quantized_matmul_kernel.py]: renamed it to quant_block_size

**Comment by lsy323** [comment]
Quantization logic looks good to me!

**Comment by vanbasten23** [comment]
Thanks for the review!
